src/main/java/com/google/api/codegen/util/ProtoParser.java

```diff
@@ -86,6 +88,57 @@ public ProtoParser(boolean enableProtoAnnotations) {
     }
   }
 
+  @Nullable
+  /* Return the name of the field representing the header parameter. */
+  public List<String> getHeaderParams(Method method) {
+    if (!enableProtoAnnotations) {
+      return null;
+    }
+
+    ImmutableSet.Builder<String> allParams = ImmutableSet.builder();
+
+    HttpRule topRule = method.getDescriptor().getMethodAnnotation(AnnotationsProto.http);
+    if (topRule == null) {
+      return null;
+    }
+
+    getHeaderParam(topRule, allParams);
+
+    // Additional bindings should only be one-deep, according to the API client config spec.
+    // No need to recurse on additional bindings' additional bindings.
+    for (HttpRule rule : topRule.getAdditionalBindingsList()) {
+      getHeaderParam(rule, allParams);
+    }
+
+    return allParams.build().asList();
+  }
+
+  // Finds the header param from a HttpRule and add the non-null value to the running set.
+  private void getHeaderParam(HttpRule httpRule, ImmutableSet.Builder<String> runningList) {
```

Comment:
`runningList` is a set. We should be rigorous here: `Set` and `List` are different data structures:
- Set: ensures uniqueness of its elements (calls `equals()` and in case of HashSet `hashCode()` of each inserted object); does not define iteration order;
- List: allows duplicates (`equals` and `hashCode` are not called); defines order of elements

Looking into how this argument is initialized, pupulated and returned (using `asList()`) I'm really confused what we really need here. Do we need just a collection, a set, a list or a list with unique elements in it (then we need to define what happens in case if a duplicate is added: the new item replaces the old one or the old one is kept)?

Reply:
We need 
- uniqueness of elements (a header param name can be repeated across different HttpRules, but we only need one of each name)
- a List return object, because that is what the calling code requires. Also, having a deterministic and easy-to-reason-about order is nice for refreshing clients and testing, so we don't have diffs with no semantic changes.

Converting an ImmutableSet to a List satisfies all those conditions.